worker/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+                    error!("{:?}", e);
+                }
+            },
+            res = child.wait() => {
+                match res {
+                    Ok(s) => {
+                        inner.finish(s).await;
+                        inner.progress.notify_waiters();
+                        return;
+                    }
+                    Err(_) => unimplemented!("when does this happen?"), // TODO
+                }
+            }
+        }
+    }
+}
+
+pub(crate) struct ProcessInner {
+    exit_status: RwLock<Option<ExitStatus>>,
+    logs: RwLock<Vec<Bytes>>,
+
+    // Signals the listeners about progress being made by the process
+    // (either new log messages or finishing).
+    progress: Arc<Notify>,
+
+    stop_sender: Mutex<Option<oneshot::Sender<()>>>,
+}
+
+impl ProcessInner {
+    fn new(stop_sender: oneshot::Sender<()>) -> Self {
+        Self {
+            exit_status: Default::default(),
+            logs: Default::default(),
+            progress: Default::default(),
+            stop_sender: Mutex::new(Some(stop_sender)),
+        }
+    }
+    async fn finish(&self, exit_status: ExitStatus) {
+        *self.exit_status.write().await = Some(exit_status);
+    }
+}
+
+/// Represents a single process.
+pub struct Process(Arc<ProcessInner>);
+
+impl Process {
+    /// Spawn a new process.
+    pub fn spawn<'a>(argv0: &str, argv: impl Iterator<Item = &'a str>) -> Result<Self, io::Error> {
+        let (stop_tx, stop_rx) = oneshot::channel();
+        let inner = Arc::new(ProcessInner::new(stop_tx));
+        let inner_clone = inner.clone();
+        let child = Command::new(argv0)
+            .args(argv)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()?;
+
+        tokio::spawn(process_task(child, inner, stop_rx));
+
+        Ok(Process(inner_clone))
+    }
+
+    /// Tries to stop the process, first gracefully by sending SIGTERM,
+    /// then, after timeout specified by `SIGTERM_TIMEOUT` elapses, by sending a SIGKILL.
+    /// If the process has already finished, returns `Ok` with the exit status.
+    /// If the process has not finished,
+    /// but another "stop" operation has already been initiated, returns `Err(())`.
+    pub async fn stop(&self) -> Result<ExitStatus, ()> {
+        if let Some(e) = *self.0.exit_status.read().await {
+            return Ok(e);
+        }
+
+        match self.0.stop_sender.lock().await.take() {
+            Some(tx) => {
+                let notify = self.0.progress.clone();
+
+                // Ignore error: if receiver has hung up, process has already finished.
+                let _ = tx.send(());
+
+                loop {
+                    let notified = notify.notified();
+                    if let Some(e) = *self.0.exit_status.read().await {
+                        return Ok(e);
+                    }
+                    notified.await;
+                }
+            }
+            None => Err(()),
```

Comment:
I think you really should try to avoid returning err here. Try to be deterministic and if we run into a race condition with another stop, either poll until it writes an exit status to the map or find another way to notify the current task when the exit status is written.

Reply:
Doing this, this should never error.

Reply:
There were some questions about this "concurrent `stop` is an error" approach during the design phase as well. https://github.com/justinas/paas/pull/1#discussion_r593497561 . So I guess its the best that I eliminate this error from both this library and the API layer.

Reply:
IIRC this is as it was defined in the design doc.

Reply:
Ah, missed that, thanks